include/boost/geometry/arithmetic/general_form.hpp

```diff
@@ -0,0 +1,161 @@
+                             geometry::get<1>(a),
+                             geometry::get<0>(b),
+                             geometry::get<1>(b));
+}
+
+template <typename T, typename Segment>
+inline general_form<T> construct_line(Segment const& segment)
+{
+    return construct_line<T>(geometry::get<0, 0>(segment),
+                             geometry::get<0, 1>(segment),
+                             geometry::get<1, 0>(segment),
+                             geometry::get<1, 1>(segment));
+}
+
+
+// Calculates intersection point of two infinite lines.
+// Returns true if the lines intersect.
+// Returns false if lines are parallel (or collinear, possibly opposite)
+template <typename Point, typename GeneralType>
+inline
+bool get_intersection(Point& ip,
+                      general_form<GeneralType> const& p,
+                      general_form<GeneralType> const& q)
+{
+    GeneralType const denominator = p.b * q.a - p.a * q.b;
+
+    static GeneralType const zero = 0;
+    if (math::equals(denominator, zero))
+    {
+        // Lines are parallel
+        return false;
+    }
+
+    // Calculate the intersection coordinates
+    geometry::set<0>(ip, (p.c * q.b - p.b * q.c) / denominator);
+    geometry::set<1>(ip, (p.a * q.c - p.c * q.a) / denominator);
+
+    return true;
+}
+
+//! Return a distance-side-measure for a point to a line
+//! Point is located left of the line if value is positive,
+//! right of the line is value is negative, and on the line if the value
+//! is exactly zero
+template <typename GeneralType, typename CoordinateType>
+inline
+typename select_most_precise<GeneralType, CoordinateType>::type
+side_value(general_form<GeneralType> const& f,
+    CoordinateType const& x, CoordinateType const& y)
+{
+    // https://en.wikipedia.org/wiki/Distance_from_a_point_to_a_line#Line_defined_by_an_equation
+    // Distance from point to line in general form is given as:
+    // (a * x + b * y + c) / sqrt(a * a + b * b);
+    // In most use cases comparisons are enough, saving the sqrt
+    // and often even the division.
+    // Also, this gives positive values for points left to the line,
+    // and negative values for points right to the line.
+    return f.a * x + f.b * y + f.c;
+}
+
+template <typename GeneralType, typename Point>
+inline
+typename select_most_precise
+<
+    GeneralType,
+    typename geometry::coordinate_type<Point>::type
+>::type
+side_value(general_form<GeneralType> const& f, Point const& p)
+{
+    return side_value(f, geometry::get<0>(p), geometry::get<1>(p));
+}
+
+// Returns true for two lines which are supposed to be (close to) collinear
+// (which is not checked) and have a similar direction
+// (in practice up to 45 degrees, TO BE VERIFIED)
+// true: -----------------> p -----------------> q
+// false: -----------------> p <----------------- q
+template <typename GeneralType>
+inline
+bool similar_direction(const general_form<GeneralType>& p,
```

Comment:
So the question is, what does "similar" mean? "Maybe up to 45 degrees"? I think the name doesn't express this well enough. Besides does such non-well-defined function is needed? If in some algorithm a check like: "angle between lines < 90 deg or < 45 deg" then this function should be named specifically expressing the desired condition and the condition in the function should check exactly what it is supposed to check.

Reply:
The documentation states: 
```
Returns true for two lines which are supposed to be (close to) collinear
(which is not checked) and have a similar direction
```
We need a function which returns true of collinear lines that are opposite, and false if they have the same direction. This simple function does the trick. But what is collinear - that is based on numerical precision. The result should be reliable with all the epsilons we use. So this function does not do any check for collinearity, and does not use any epsilon. Its implementation is straightforward, and (basically as a side effect) it returns true for lines even 30 degrees apart. Which indeed have a `similar` direction (therefore I still like the name, it should not be too precise)

Reply:
> We need a function which returns true of collinear lines that are opposite, and false if they have the same direction. 

So the precondition is that the arguments are two collinear lines (so this is checked before somewhere) and then they either may be going into the same direction or the opposite and you have to check that, correct? If that's correct then this function has strictly defined purpose. Not to check similarity but to check the direction of two collinear lines. So why not name it to something like: `same_collinear_direction` or since in the code we call this being opposite why not simply: `are_opposite` or `check_opposite`?

Reply:
I don't know why `same_collinear_direction` is more clear than `similar_direction`. With `are_opposite` I have the objection that it does not check collinearity itself. 
But I'm not opposed to renaming ;-)

Reply:
I agree, the names I proposed are also not perfect. If you think of anything better then rename it. If not I'm ok with `similar_direction`.

Reply:
Thank you, for now I will merge it - maybe we have more ideas later.